From 23c1e2fb5b8607b0c7926ac1b0cf40ee125f7c05 Mon Sep 17 00:00:00 2001
From: Ted Zlatanov <tzz@lifelogs.com>
Date: Mon, 27 Jul 2015 16:02:03 -0400
Subject: [PATCH] Redmine#7413: minor fix for some overzealous lexers

Changelog: Fix "@endif" keyword sometimes being improperly processed
by policy parser.
---
 libpromises/cf3lex.l                           |  8 +++----
 tests/acceptance/00_basics/macros/if_ignore.cf | 30 ++++++++++++++++++++++++++
 2 files changed, 34 insertions(+), 4 deletions(-)
 create mode 100644 tests/acceptance/00_basics/macros/if_ignore.cf

diff --git a/libpromises/cf3lex.l b/libpromises/cf3lex.l
index 54f6bf7..dd93209 100644
--- a/libpromises/cf3lex.l
+++ b/libpromises/cf3lex.l
@@ -210,9 +210,9 @@ promise_type   [a-zA-Z_]+:
                                  P.if_depth--;
                                }
 
-<if_ignore_state>.*          {
+<if_ignore_state>[^@]      {
                                ParserDebug("\tL:inside macro @if, ignoring text:%s\n", yytext);
-                             }
+                           }
 
 {promises}            {
                           /* Note this has to come before "id" since it is a subset of id */
@@ -436,7 +436,7 @@ promise_type   [a-zA-Z_]+:
 <if_ignore_state><<EOF>>     {
                                if (P.if_depth > 0)
                                {
-                                 yyerror("EOF seen while @if was waiting for @endif");
+                                 yyerror("EOF seen while @if was waiting for @endif in ignore_state");
                                  return 0;
                                }
                              }
@@ -444,7 +444,7 @@ promise_type   [a-zA-Z_]+:
 <<EOF>>     {
               if (P.if_depth > 0)
               {
-                yyerror("EOF seen while @if was waiting for @endif");
+                yyerror("EOF seen while @if was waiting for @endif without ignore_state");
               }
               return 0; // loops forever without this
             }
diff --git a/tests/acceptance/00_basics/macros/if_ignore.cf b/tests/acceptance/00_basics/macros/if_ignore.cf
new file mode 100644
index 0000000..108585e
--- /dev/null
+++ b/tests/acceptance/00_basics/macros/if_ignore.cf
@@ -0,0 +1,30 @@
+######################################################
+#
+#  Test that @if works for greater versions
+#
+#####################################################
+
+body common control
+{
+      inputs => { "../../default.cf.sub" };
+      bundlesequence  => { default("$(this.promise_filename)") };
+      version => "1.0";
+}
+
+bundle common test
+{
+@if minimum_version(300.600)
+      Some new function here...
+
+
+      more text...
+@endif
+}
+
+bundle agent check
+{
+  methods:
+      "" usebundle => dcs_passif_expected("cfengine",
+                                         "",
+                                         $(this.promise_filename));
+}
